ci/kind/kind-setup.sh

```diff
@@ -517,18 +517,20 @@ function destroy_external_servers {
 
 function clean_kind {
     echo "=== Cleaning up stale kind clusters ==="
-    read -a all_kind_clusters <<< $(kind get clusters)
-    for kind_cluster_name in "${all_kind_clusters[@]}"; do
-        creationTimestamp=$(kubectl get nodes --context kind-$kind_cluster_name -o json -l node-role.kubernetes.io/control-plane | \
+    for context in $(kubectl config get-contexts -o name | grep 'kind-'); do
+      cluster_name=$(echo $context | sed 's/^kind-//')
+      if docker ps --format '{{.Names}}' | grep -q "$cluster_name"; then
```

Comment:
why are we relying on this docker command instead of `kind get clusters`?

Reply:
And `kind get nodes` can also be used to list all node names.

Reply:
We are not relying on kind get clusters because it lists those clusters also for which the creation is not yet completed or successful, and is in mid way, which creates a problem because when we try to get contexts of such clusters which are yet not completed it returns an error. 
This error we may encounter in case of aborted jobs and multiple jobs run.

In case of aborted job suppose we abort the job as soon as the cluster creation starts, so when we do `kind get clusters` we will get to see that cluster from aborted job in the list but since its context will not be available so the cleanup will panic and job will fail.

In case of multiple jobs run suppose there are two jobs running parallely and one job has just triggered cluster creation step and the other job triggers the cleanup function so it will list the cluster which is ye not created and then it will try to fetch the context for that cluster and the job will fail because of panic.

Reply:
Sometimes because of these stale clusters present in the environment testbed also becomes unhealthy.

Reply:
@jainpulkit22 Thanks for the explanation, my concern is whether `docker ps --format '{{.Names}}' | grep -q "$cluster_name"` is a sufficient basis to determine if a cluster is ready. Is there something like status.conditions fields in the context that would allow us to accurately determine if the cluster is ready?

Reply:
I was trying to point out that at this stage we are already iterating over known contexts (`for context in $(kubectl config get-contexts -o name | grep 'kind-')`, so we know that the context exists. It is not the same as the previous solution, which was first calling `kind get clusters`, and then calling `kubectl config get-contexts`.

In your comment and in the original issue, you talk about a "panic", but it's not clear to me what failure you are referring to exactly.

Note that even with your proposed solution, races seem possible:
1) job 1 calls `kubectl config get-contexts` and gets `context1`
2) job 2 calls `kubectl config get-contexts` and gets `context1`
3) job 1 calls `docker ps` and finds the container it is looking for
4) job 2 calls `docker ps` and finds the container it is looking for
5) job 1 calls `kubectl get nodes --context context1 ...` succesfully
6) job 1 deletes the kind cluster sucessfully
7) job 2 calls `kubectl get nodes --context context1 ...` but `context1` does not exist anymore!

Either the code should be best effort and tolerate failures, or maybe you should use a lock (see `flock`) for correctness, and then you don't need to worry about concurrent executions and you can use the most appropriate commands for the task.

Reply:
>In your comment and in the original issue, you talk about a "panic", but it's not clear to me what failure you are referring to exactly.

The failure I talked about in the issue can be explained with the following example.
Suppose there are two jobs running in parallel JobA and JobB, now one of the jobs has just triggered the cluster creation, and the other job say JobB in in the initial cleanup phase, so when it will run the `kind get clusters` command it will list the cluster for which the creation has just started by `JobA`, and then when it tries to run the command `kubectl get nodes --context ...` it will panic because the context does not exist or we can say that cluster creation is yet not completed.

Now, coming to the race condition which you mentioned, I think that this race condition will not be there in practical, because there will be few seconds of time lag between two parallel runs so if one has already deleted the cluster then docker ps will not list that container.
So, the race condition which you mentioned here seems to be a race condition in Ideal scenario but not practically possible.

Reply:
I still believe that the `docker` command is unnecessary:

> We are not relying on kind get clusters because it lists those clusters also for which the creation is not yet completed or successful, and is in mid way, which creates a problem because when we try to get contexts of such clusters which are yet not completed it returns an error.

The outer loop is still `for context in $(kubectl config get-contexts -o name | grep 'kind-'); do`, so we already have the context

> it will panic 

Do you mean a Go panic in `kubectl`?

> So, the race condition which you mentioned here seems to be a race condition in Ideal scenario but not practically possible.

2 jobs can be triggered at the same time. We should probably be tolerant of failures. If the `kubectl` command (`kubectl get nodes`) fails or the `kind` command fails (`kind delete cluster --name $cluster_name`), we don't have to fail the script.

Reply:
> I still believe that the `docker` command is unnecessary:
The docker command is necessary here because sometimes the contexts are there in the kubectl config get contexts but the cluster is deleted or is in the creation step.

> The outer loop is still `for context in $(kubectl config get-contexts -o name | grep 'kind-'); do`, so we already have the context
Yes just before the completion the context is populated.

> Do you mean a Go panic in `kubectl`?
Yes

> 2 jobs can be triggered at the same time. We should probably be tolerant of failures. If the `kubectl` command (`kubectl get nodes`) fails or the `kind` command fails (`kind delete cluster --name $cluster_name`), we don't have to fail the script.
Okay we can have a fault tolerance for this.

Reply:
Updated the code, this ensures fault tolerance, I haven't used flock because i guess that would affect he parallel job run because if some job is in cluster creation or is running e2e and the oter has acuired floack over the script file then it will not be able to read further data.